far/vmenu.cpp

```diff
@@ -2129,253 +2365,139 @@ void VMenu::ShowMenu(bool IsParent)
-				}
+	const auto [TextBegin, TextWidth] { Layout.Text.value() };
 
-				if (!Items[I].Annotations.empty())
-				{
-						size_t Pos = 0;
-						for (const auto& [AnnPos, AnnSize]: Items[I].Annotations)
-						{
-							const int StartOffset = 1; // 1 is '<<' placeholder size
-							const size_t pre_len = AnnPos - Items[I].ShowPos + StartOffset - Pos + 1;
-							if (Pos < strMenuLine.size())
-							{
-								Text(string_view(strMenuLine).substr(Pos, pre_len));
-								Pos += pre_len;
-								if (Pos < strMenuLine.size())
-								{
-									SetColor(Col);
-									Text(string_view(strMenuLine).substr(Pos, AnnSize));
-									Pos += AnnSize;
-									SetColor(CurColor);
-								}
-							}
-						}
-						if (Pos < strMenuLine.size())
-							Text(string_view(strMenuLine).substr(Pos));
-				}
-				else
-				{
-					if (HotkeyVisualPos != string::npos || Items[I].AutoHotkey)
-					{
-						const auto HotkeyPos = (HotkeyVisualPos != string::npos? HotkeyVisualPos : Items[I].AutoHotkeyPos) + PrefixSize;
-
-						Text(string_view(strMenuLine).substr(0, HotkeyPos));
-
-						if (HotkeyPos < strMenuLine.size())
-						{
-							const auto SaveColor = CurColor;
-							SetColor(Col);
-							Text(strMenuLine[HotkeyPos]);
-							SetColor(SaveColor);
-							Text(string_view(strMenuLine).substr(HotkeyPos + 1));
-						}
-					}
-					else
-					{
-						Text(strMenuLine);
-					}
-				}
+	const item_color_indicies ColorIndices{ CurItem };
+	auto CurColorIndex{ ColorIndices.Normal };
+	auto AltColorIndex{ ColorIndices.Highlighted };
 
-				// сделаем добавочку для NO_BOX
-				{
-					int Width = m_Where.right - WhereX() + (m_BoxType == NO_BOX? 1 : 0);
-					if (Width > 0)
-						Text(string(Width, L' '));
-				}
+	auto CurPos{ static_cast<int>(CurItem.ShowPos) };
+	const auto EndPos{ std::min(CurPos + TextWidth, ItemTextSize) };
 
-				if (Items[I].Flags & MIF_SUBMENU)
-				{
-					GotoXY(m_Where.right - 1, Y);
-					Text(L'►');
-				}
+	HighlightMarkup.clear();
+	if (!CurItem.Annotations.empty())
+	{
+		MarkupSliceBoundaries(
+			std::pair{ CurPos, EndPos },
+			CurItem.Annotations | std::views::transform([](const auto Ann) { return std::pair{ Ann.first, Ann.first + Ann.second }; }),
+			HighlightMarkup);
+	}
+	else if (HotkeyPos != string::npos || CurItem.AutoHotkey)
+	{
+		const auto HighlightPos = static_cast<int>(HotkeyPos != string::npos? HotkeyPos : CurItem.AutoHotkeyPos);
+		MarkupSliceBoundaries(
+			std::pair{ CurPos, EndPos },
+			std::views::single(std::pair{ HighlightPos, HighlightPos + 1 }),
+			HighlightMarkup);
+	}
+	else
+	{
+		HighlightMarkup.emplace_back(EndPos);
+	}
 
-				SetColor(Colors[(Items[I].Flags&LIF_DISABLE)?VMenuColorArrowsDisabled:(Items[I].Flags&LIF_SELECTED?VMenuColorArrowsSelect:VMenuColorArrows)]);
+	GotoXY(TextBegin, Y);
+	for (const auto SliceEnd : HighlightMarkup)
+	{
+		SetColor(Colors, CurColorIndex);
+		Text(string_view{ ItemTextToDisplay.begin() + CurPos, ItemTextToDisplay.begin() + SliceEnd });
```

Comment:
`string_view(ItemTextToDisplay).substr(CurPos, SliceEnd - CurPos)` maybe?

Reply:
I thought about it. No real preference. Let's do it your way.